File: walczak/wri-fish.py
import numpy as np

# Population size
N=1000

# Maximum number of generations
Tmax=1000

# Number of simulations
nsims=1000

# Function to compute the population that has the mutation after a number of
# generations
def end_population(init_n,s):

    # Initial population that has the mutation
    n=init_n

    # Simulate a number of generations
    for j in xrange(Tmax):
        p=float(n)/N*(1+s)/(1.+s*n/N)
        n=np.random.binomial(N,p,1)[0]

        # Check if we have either zero or N individuals. In either case, we can
        # terminate early
        if n==0 or n==N:
            break
    return n

# Test a range of different fitnesses
for s in np.linspace(0,1,41):

    # Initialize counters
    nfix=0;nlose=0

    # Initialize the number of individuals that have the mutation
    init_n=1
    #init_n=int(round((1.+s)/((1./N)*s+1)))
    
    # Perform a number of trials with this fitness
    for i in xrange(1000):
        e=end_population(init_n,s)
        if e==0:
            nlose+=1
        elif e==N:
            nfix+=1

    # Print the proportion of cases where the mutation fixed 
    print s,init_n,nlose,nfix,float(nfix)/(nfix+nlose)

File: walczak/wri-fish.cc
#include "wri-fish.hh"

#include <gsl/gsl_randist.h>

/** The class constructor sets constants and initializes the GSL random number
 * generator.
 * \param[in] N_ the total population size.
 * \param[in] Tmax_ the maximum number of generations.
 * \param[in] seed a seed for the random number generator. */
wright_fisher::wright_fisher(int N_,int Tmax_,int seed) : N(N_), Ninv(1./N),
    Tmax(Tmax_), rng(gsl_rng_alloc(gsl_rng_taus)) {
    gsl_rng_set(rng,seed);
}

/** The class destructor frees the dynamically allocated random number
 * generator. */
wright_fisher::~wright_fisher() {
    gsl_rng_free(rng);
}

/** Computes the end population of individuals with a fitness of s.
 * \param[in] init_n the initial number of individuals in this population.
 * \param[in] s the fitness.
 * \return The end population. */
int wright_fisher::end_population(int init_n,double s) {
    int n=init_n;
    double p;

    // Perform the generations
    for(int i=0;i<Tmax;i++) {
        p=Ninv*(1+s)*n/(1.+s*Ninv*n);
        n=gsl_ran_binomial(rng,p,N);
        if(n==0||n==N) break;
    }
    return n;
}
